packages/adblocker-content/adblocker.ts

```diff
@@ -70,41 +87,41 @@ export function extractFeaturesFromDOM(roots: Element[]): {
 } {
   // NOTE: This cannot be global as puppeteer needs to be able to serialize this function.
   const ignoredTags = new Set(['br', 'head', 'link', 'meta', 'script', 'style', 's']);
-  const classes: Set<string> = new Set();
-  const hrefs: Set<string> = new Set();
-  const ids: Set<string> = new Set();
+  // Try to reduce duplicate entries by using Set instead of an array.
+  const elements: Set<Element> = new Set();
 
   for (const root of roots) {
-    for (const element of [
-      root,
-      ...root.querySelectorAll(
-        '[id]:not(html):not(body),[class]:not(html):not(body),[href]:not(html):not(body)',
-      ),
-    ]) {
+    elements.add(root);
+
+    for (const element of root.querySelectorAll(
+      '[id]:not(html):not(body),[class]:not(html):not(body),[href]:not(html):not(body)',
+    )) {
       if (ignoredTags.has(element.nodeName.toLowerCase())) {
         continue;
       }
 
-      // Update ids
-      const id = element.id;
-      if (id) {
-        ids.add(id);
-      }
+      elements.add(element);
+    }
+  }
 
-      // Update classes
-      const classList = element.classList;
-      if (classList) {
-        for (const cls of classList) {
-          classes.add(cls);
-        }
-      }
+  const classes: Set<string> = new Set();
+  const hrefs: Set<string> = new Set();
+  const ids: Set<string> = new Set();
+
+  for (const element of elements) {
+    if (element.id) {
+      ids.add(element.id);
+    }
 
-      // Update href
-      const href = element.getAttribute('href');
-      if (typeof href === 'string') {
-        hrefs.add(href);
+    if (element.classList) {
```

Comment:
the classList is probably always there. also, this is probably not faster than before. please revert